### PR TITLE
Design Document for SRE Challenge

### PR DESCRIPTION
**Pull request by dereknorrbom**
Design Document for SRE Challenge

### CODE REVIEW

File: rfd/0000-sre-challenge.md
Comment (evanfreed) [inline rfd/0000-sre-challenge.md]: Any reason against using TLS 1.3?

File: rfd/0000-sre-challenge.md
Comment (evanfreed) [inline rfd/0000-sre-challenge.md]: How will the API support specifying the namespace?

File: rfd/0000-sre-challenge.md
Comment (evanfreed) [inline rfd/0000-sre-challenge.md]: Just clarifying since you asked about this in the Slack channel.  Is this something you plan on implementing? If not, maybe include a comment that this is how you would go about this.

File: rfd/0000-sre-challenge.md
Comment (taraspos) [inline rfd/0000-sre-challenge.md]: Similar to the question below:
- How are we going to provide provide deployment name in the request?
- What kind of output should we expect?

File: rfd/0000-sre-challenge.md
Comment (taraspos) [inline rfd/0000-sre-challenge.md:142]: `clean` is fairy common command to run and not sure if I would want it to destroy the cluster every time.
What do you think about separate command something like `teardown`?

File: rfd/0000-sre-challenge.md
Comment (taraspos) [inline rfd/0000-sre-challenge.md]: What kind of OS will your script target?

Installation for those tools might be slightly different for MacOS/Linux/Windows

File: rfd/0000-sre-challenge.md
Comment (taraspos) [inline rfd/0000-sre-challenge.md:215]: What kind of kubernetes resources are you planning to deploy with the helm chart?

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: I've been fortunate enough to work in environments where I only needed to set the TLS version without memorizing all the ciphers. I initially went with TLS 1.2 based on common recommendations. However, TLS 1.3 is the better choice for stronger security and faster performance. I'll update the design document to specify TLS 1.3.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: I planned to hardcode the namespace value based on the challenge documentation's suggestion to "_Use hardcoded values as much as possible and simply add TODO items showing your thinking._"  To support specifying the namespace, I can modify the API design to accept it as a query parameter.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: I do not plan on supporting multiple clusters within the scope of this project. I will update the document to make it clear that the section on multiple clusters is just an explanation of how I could support it if it were part of the project scope.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: I planned to hardcode the deployment name based on the challenge documentation's suggestion to "_Use hardcoded values as much as possible and simply add TODO items showing your thinking._" To support specifying the deployment name, I will modify the API design to accept it as a query parameter.  

The API will return JSON responses for all endpoints, I will update the design document to reflect this.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:142]: I agree that having the clean command destroy the cluster may not be ideal. I'll separate the cluster and Docker teardown into a different command.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: The setup.sh script will target x86 64-bit Linux and MacOS environments, as specified in the challenge documentation.  Windows is out of scope for this project, so the script will not include Windows-specific installation instructions.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:215]: The Helm chart will deploy the following Kubernetes resources:

1. Deployment: This will manage the deployment of the API Docker container, ensuring the specified number of replicas are running.
2. Service: This will expose the API deployment within the Kubernetes cluster, allowing it to be accessed by other services or through port forwarding.
3. ClusterRole and ClusterRoleBinding: These will grant cluster-wide permissions necessary for the API to interact with Kubernetes resources.
4. Role and RoleBinding: These will grant namespace-specific permissions for the API to interact with Kubernetes resources.

**Comment by dereknorrbom** [comment]
The design document has been updated to reflect the comment questions/feedback. @evanfreed @taraspos

File: rfd/0000-sre-challenge.md
Comment (jimbishopp) [inline rfd/0000-sre-challenge.md]: These two endpoints are very similar. They both accept the same parameters and operate on replica count. What is the reason for defining two endpoints rather than one endpoint with different HTTP verbs?

File: rfd/0000-sre-challenge.md
Comment (jimbishopp) [inline rfd/0000-sre-challenge.md]: The response message reads a bit like a log message. What do you think about returning the same message as returned from the get replica count endpoint?

File: rfd/0000-sre-challenge.md
Comment (jimbishopp) [inline rfd/0000-sre-challenge.md]: Will the deployment strings in the array include namespace?

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: In pedantic REST, the endpoints would use the same name with different verbs. I went with descriptive names for clarity.

Proposed Update:
- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
- **POST /replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: Returning the same message as the GET replica count endpoint makes sense for consistency. I'll update the response to match the GET endpoint.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md]: Currently, the design does not include the namespace in the deployment strings. Including the namespace can provide more clarity, especially when working with multiple namespaces.

Proposed update:

```json
{
  "deployments": ["default/my-deployment"]
}
```

**Comment by dereknorrbom** [comment]
The design document has been updated to reflect the comment questions/feedback. @jimbishopp @evanfreed @taraspos

File: rfd/0000-sre-challenge.md
Comment (logand22) [inline rfd/0000-sre-challenge.md:23]: Are these query parameters required?

File: rfd/0000-sre-challenge.md
Comment (logand22) [inline rfd/0000-sre-challenge.md:29]: Can you provide some examples of error responses.

File: rfd/0000-sre-challenge.md
Comment (logand22) [inline rfd/0000-sre-challenge.md:145]: But where do these certs originate from? Are they built into the container? If so can you provide more information on this.

File: rfd/0000-sre-challenge.md
Comment (logand22) [inline rfd/0000-sre-challenge.md:176]: Just to clarify, you'll create an in-memory cache and then use `informers` to populate that cache? If so, is there a reason you don't use the cache that's built into the informer?

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:23]: They were hardcoded originally per "_Use hardcoded values as much as possible and simply add TODO items showing your thinking_", and updated to utilize query parameters based on this comment https://github.com/dereknorrbom/k8s-deployment-scaler/pull/1#discussion_r1644183083.  I plan on using the hardcoded values as fallbacks when a deployment/namespace is not provided in the request.

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:145]: For simplicity, the certs are self-signed and generated within the container.

File: rfd/0000-sre-challenge.md
Comment (logand22) [reply rfd/0000-sre-challenge.md:23]: Sorry, I meant can you leave the namespace and deployment parameters empty and receive all replica counts? Or are you saying your application will only support monitoring a single deployment?

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:176]: It was an oversight on my part. I'll update the design to use the informer's built-in cache.

**Review by logand22 — APPROVED** [review]

**Review by evanfreed — APPROVED** [review]
@dereknorrbom since you have the approvals necessary feel free to merge this when ready and move onto the implementation part of the challenge. 

As a reminder please focus on the `Server` components for the first PR (if there's some simple build targets to support it that's fine).

File: rfd/0000-sre-challenge.md
Comment (dereknorrbom) [reply rfd/0000-sre-challenge.md:29]: Yes, I will update the design document to include example error responses.

**Review by taraspos — APPROVED** [review]

**Pull request merged**
